In [1]:
import numpy as np
import pandas as pd
import statsmodels.api as sm

# Reading the Data

In [2]:
data = pd.read_csv('general_data.csv')

In [4]:
data.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   51        No      Travel_Rarely                   Sales                 6   
1   31       Yes  Travel_Frequently  Research & Development                10   
2   32        No  Travel_Frequently  Research & Development                17   
3   38        No         Non-Travel  Research & Development                 2   
4   32        No      Travel_Rarely  Research & Development                10   

   Education EducationField  EmployeeCount  EmployeeID  Gender  ...  \
0          2  Life Sciences              1           1  Female  ...   
1          1  Life Sciences              1           2  Female  ...   
2          4          Other              1           3    Male  ...   
3          5  Life Sciences              1           4    Male  ...   
4          1        Medical              1           5    Male  ...   

   NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0                 1.0      Y                11              8   
1                 0.0      Y                23              8   
2                 1.0      Y                15              8   
3                 3.0      Y                11              8   
4                 4.0      Y                12              8   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 0               1.0                      6               1   
1                 1               6.0                      3               5   
2                 3               5.0                      2               5   
3                 3              13.0                      5               8   
4                 2               9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  

[5 rows x 24 columns]

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4410 entries, 0 to 4409
Data columns (total 24 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Age                      4410 non-null   int64  
 1   Attrition                4410 non-null   object 
 2   BusinessTravel           4410 non-null   object 
 3   Department               4410 non-null   object 
 4   DistanceFromHome         4410 non-null   int64  
 5   Education                4410 non-null   int64  
 6   EducationField           4410 non-null   object 
 7   EmployeeCount            4410 non-null   int64  
 8   EmployeeID               4410 non-null   int64  
 9   Gender                   4410 non-null   object 
 10  JobLevel                 4410 non-null   int64  
 11  JobRole                  4410 non-null   object 
 12  MaritalStatus            4410 non-null   object 
 13  MonthlyIncome            4410 non-null   int64  
 14  NumCompaniesWorked      

In [6]:
data.describe()

Age  DistanceFromHome    Education  EmployeeCount   EmployeeID  \
count  4410.000000       4410.000000  4410.000000         4410.0  4410.000000   
mean     36.923810          9.192517     2.912925            1.0  2205.500000   
std       9.133301          8.105026     1.023933            0.0  1273.201673   
min      18.000000          1.000000     1.000000            1.0     1.000000   
25%      30.000000          2.000000     2.000000            1.0  1103.250000   
50%      36.000000          7.000000     3.000000            1.0  2205.500000   
75%      43.000000         14.000000     4.000000            1.0  3307.750000   
max      60.000000         29.000000     5.000000            1.0  4410.000000   

          JobLevel  MonthlyIncome  NumCompaniesWorked  PercentSalaryHike  \
count  4410.000000    4410.000000         4391.000000        4410.000000   
mean      2.063946   65029.312925            2.694830          15.209524   
std       1.106689   47068.888559            2.498887           3.659108   
min       1.000000   10090.000000            0.000000          11.000000   
25%       1.000000   29110.000000            1.000000          12.000000   
50%       2.000000   49190.000000            2.000000          14.000000   
75%       3.000000   83800.000000            4.000000          18.000000   
max       5.000000  199990.000000            9.000000          25.000000   

       StandardHours  StockOptionLevel  TotalWorkingYears  \
count         4410.0       4410.000000        4401.000000   
mean             8.0          0.793878          11.279936   
std              0.0          0.851883           7.782222   
min              8.0          0.000000           0.000000   
25%              8.0          0.000000           6.000000   
50%              8.0          1.000000          10.000000   
75%              8.0          1.000000          15.000000   
max              8.0          3.000000          40.000000   

       TrainingTimesLastYear  YearsAtCompany  YearsSinceLastPromotion  \
count            4410.000000     4410.000000              4410.000000   
mean                2.799320        7.008163                 2.187755   
std                 1.288978        6.125135                 3.221699   
min                 0.000000        0.000000                 0.000000   
25%                 2.000000        3.000000                 0.000000   
50%                 3.000000        5.000000                 1.000000   
75%                 3.000000        9.000000                 3.000000   
max                 6.000000       40.000000                15.000000   

       YearsWithCurrManager  
count           4410.000000  
mean               4.123129  
std                3.567327  
min                0.000000  
25%                2.000000  
50%                3.000000  
75%                7.000000  
max               17.000000

# Cleaning the Data

In [16]:
data.isnull().any()

Age                        False
Attrition                  False
BusinessTravel             False
Department                 False
DistanceFromHome           False
Education                  False
EducationField             False
EmployeeCount              False
EmployeeID                 False
Gender                     False
JobLevel                   False
JobRole                    False
MaritalStatus              False
MonthlyIncome              False
NumCompaniesWorked         False
Over18                     False
PercentSalaryHike          False
StandardHours              False
StockOptionLevel           False
TotalWorkingYears          False
TrainingTimesLastYear      False
YearsAtCompany             False
YearsSinceLastPromotion    False
YearsWithCurrManager       False
dtype: bool

In [17]:
data.fillna(0 , inplace=True)

In [18]:
data.isnull().any()

Age                        False
Attrition                  False
BusinessTravel             False
Department                 False
DistanceFromHome           False
Education                  False
EducationField             False
EmployeeCount              False
EmployeeID                 False
Gender                     False
JobLevel                   False
JobRole                    False
MaritalStatus              False
MonthlyIncome              False
NumCompaniesWorked         False
Over18                     False
PercentSalaryHike          False
StandardHours              False
StockOptionLevel           False
TotalWorkingYears          False
TrainingTimesLastYear      False
YearsAtCompany             False
YearsSinceLastPromotion    False
YearsWithCurrManager       False
dtype: bool

In [19]:
data.drop_duplicates()

Age  Attrition     BusinessTravel              Department  \
0      51          0      Travel_Rarely                   Sales   
1      31          1  Travel_Frequently  Research & Development   
2      32          0  Travel_Frequently  Research & Development   
3      38          0         Non-Travel  Research & Development   
4      32          0      Travel_Rarely  Research & Development   
...   ...        ...                ...                     ...   
4405   42          0      Travel_Rarely  Research & Development   
4406   29          0      Travel_Rarely  Research & Development   
4407   25          0      Travel_Rarely  Research & Development   
4408   42          0      Travel_Rarely                   Sales   
4409   40          0      Travel_Rarely  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount  EmployeeID  \
0                    6          2  Life Sciences              1           1   
1                   10          1  Life Sciences              1           2   
2                   17          4          Other              1           3   
3                    2          5  Life Sciences              1           4   
4                   10          1        Medical              1           5   
...                ...        ...            ...            ...         ...   
4405                 5          4        Medical              1        4406   
4406                 2          4        Medical              1        4407   
4407                25          2  Life Sciences              1        4408   
4408                18          2        Medical              1        4409   
4409                28          3        Medical              1        4410   

      Gender  ...  NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0     Female  ...                 1.0      1                11              8   
1     Female  ...                 0.0      1                23              8   
2       Male  ...                 1.0      1                15              8   
3       Male  ...                 3.0      1                11              8   
4       Male  ...                 4.0      1                12              8   
...      ...  ...                 ...    ...               ...            ...   
4405  Female  ...                 3.0      1                17              8   
4406    Male  ...                 2.0      1                15              8   
4407    Male  ...                 0.0      1                20              8   
4408    Male  ...                 0.0      1                14              8   
4409    Male  ...                 0.0      1                12              8   

      StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                    0                1.0                      6   
1                    1                6.0                      3   
2                    3                5.0                      2   
3                    3               13.0                      5   
4                    2                9.0                      2   
...                ...                ...                    ...   
4405                 1               10.0                      5   
4406                 0               10.0                      2   
4407                 0                5.0                      4   
4408                 1               10.0                      2   
4409                 0                0.0                      6   

      YearsAtCompany  YearsSinceLastPromotion  YearsWithCurrManager  
0                  1                        0                     0  
1                  5                        1                     4  
2                  5                        0                     3  
3                  8                        7                     5  
4                  6                        0                     4  
...              ...                      

# Converting the Data

In [20]:
data.Attrition = data.Attrition.replace('No', 0)
data.Attrition = data.Attrition.replace('Yes', 1)

In [21]:
data.Over18 = data.Over18.replace('N', 0)
data.Over18 = data.Over18.replace('Y', 1)

In [22]:
data.head()

Age  Attrition     BusinessTravel              Department  \
0   51          0      Travel_Rarely                   Sales   
1   31          1  Travel_Frequently  Research & Development   
2   32          0  Travel_Frequently  Research & Development   
3   38          0         Non-Travel  Research & Development   
4   32          0      Travel_Rarely  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeID  \
0                 6          2  Life Sciences              1           1   
1                10          1  Life Sciences              1           2   
2                17          4          Other              1           3   
3                 2          5  Life Sciences              1           4   
4                10          1        Medical              1           5   

   Gender  ...  NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0  Female  ...                 1.0      1                11              8   
1  Female  ...                 0.0      1                23              8   
2    Male  ...                 1.0      1                15              8   
3    Male  ...                 3.0      1                11              8   
4    Male  ...                 4.0      1                12              8   

   StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 0                1.0                      6               1   
1                 1                6.0                      3               5   
2                 3                5.0                      2               5   
3                 3               13.0                      5               8   
4                 2                9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  

[5 rows x 24 columns]

# Loading X and Y values

In [51]:
y = data.Attrition
y

0       0
1       1
2       0
3       0
4       0
       ..
4405    0
4406    0
4407    0
4408    0
4409    0
Name: Attrition, Length: 4410, dtype: int64

In [52]:
x = data[['Age','DistanceFromHome','Education','JobLevel','MonthlyIncome','NumCompaniesWorked','PercentSalaryHike','TotalWorkingYears','TrainingTimesLastYear','YearsAtCompany','YearsWithCurrManager']]

In [53]:
x

Age  DistanceFromHome  Education  JobLevel  MonthlyIncome  \
0      51                 6          2         1         131160   
1      31                10          1         1          41890   
2      32                17          4         4         193280   
3      38                 2          5         3          83210   
4      32                10          1         1          23420   
...   ...               ...        ...       ...            ...   
4405   42                 5          4         1          60290   
4406   29                 2          4         1          26790   
4407   25                25          2         2          37020   
4408   42                18          2         1          23980   
4409   40                28          3         2          54680   

      NumCompaniesWorked  PercentSalaryHike  TotalWorkingYears  \
0                    1.0                 11                1.0   
1                    0.0                 23                6.0   
2                    1.0                 15                5.0   
3                    3.0                 11               13.0   
4                    4.0                 12                9.0   
...                  ...                ...                ...   
4405                 3.0                 17               10.0   
4406                 2.0                 15               10.0   
4407                 0.0                 20                5.0   
4408                 0.0                 14               10.0   
4409                 0.0                 12                0.0   

      TrainingTimesLastYear  YearsAtCompany  YearsWithCurrManager  
0                         6               1                     0  
1                         3               5                     4  
2                         2               5                     3  
3                         5               8                     5  
4                         2               6                     4  
...                     ...             ...                   ...  
4405                      5               3                     2  
4406                      2               3                     2  
4407                      4               4                     2  
4408                      2               9                     8  
4409                      6              21                     9  

[4410 rows x 11 columns]

In [54]:
x1 = sm.add_constant(x) 

In [55]:
Logistic = sm.Logit(y,x1)

In [56]:
result=Logistic.fit()

Optimization terminated successfully.
         Current function value: 0.411260
         Iterations 6


# Result

In [58]:
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:              Attrition   No. Observations:                 4410
Model:                          Logit   Df Residuals:                     4398
Method:                           MLE   Df Model:                           11
Date:                Sun, 09 Aug 2020   Pseudo R-squ.:                 0.06890
Time:                        11:38:52   Log-Likelihood:                -1813.7
converged:                       True   LL-Null:                       -1947.9
Covariance Type:            nonrobust   LLR p-value:                 3.813e-51
=========================================================================================
                            coef    std err          z      P>|z|      [0.025      0.975]
-----------------------------------------------------------------------------------------
const                     0.4780      0.335      1.428      0.153      -0.178       1.134
Age                      -0.0345      0.007     -5.150      0.000      -0.048      -0.021
DistanceFromHome         -0.0028      0.005     -0.537      0.591      -0.013       0.007
Education                -0.0476      0.041     -1.149      0.251      -0.129       0.034
JobLevel                 -0.0373      0.039     -0.960      0.337      -0.113       0.039
MonthlyIncome         -1.912e-06   9.35e-07     -2.045      0.041   -3.74e-06   -7.97e-08
NumCompaniesWorked        0.1081      0.018      6.049      0.000       0.073       0.143
PercentSalaryHike         0.0125      0.011      1.099      0.272      -0.010       0.035
TotalWorkingYears        -0.0545      0.012     -4.725      0.000      -0.077      -0.032
TrainingTimesLastYear    -0.1260      0.034     -3.693      0.000      -0.193      -0.059
YearsAtCompany            0.0475      0.016      2.976      0.003       0.016       0.079
YearsWithCurrManager     -0.1238      0.022     -5.736      0.000      -0.166      -0.082
=========================================================================================
"""

Hence we can say that

__P(Attrition = Yes/No) = 1 / (1+e^-(0.4780 - 0.0345(Age) - 0.0028(DistanceFromHome) - 0.0476(Education) - 0.0373(JobLevel) - 1.912e-06(MonthlyIncome) + 0.1081(NumCompaniesWorked) + 0.0125(PercentSalaryHike) - 0.0545(TotalWorkingYears) - 0.1260(TrainingTimesLastYear) + 0.0475(YearsAtCompany) - 0.1238(YearsWithCurrManager)))__ 